src/psyclone/tests/psyir/transformations/replace_induction_variables_test.py

```diff
@@ -0,0 +1,266 @@
+            "3 * SIN(i - 1))" in out_all)
+    assert "a(13) = 7 + (13 + 1) * 13" in out_loop
+    # Make sure the assignment to t1%a has been added outside of the loop:
+    assert "t1%a = 13" in out_all
+    assert "t1%a = 13" not in out_loop
+
+
+# ----------------------------------------------------------------------------
+def test_riv_not_movable(fortran_reader, fortran_writer):
+    '''Tests assignments that cannot be moved.
+    '''
+    source = '''program test
+                integer i, ic1, ic2, ic3, ic4, ic5, ic6
+                integer :: some_func
+                real, dimension(10) :: a
+                do i = 1, 10
+                    ic1 = i+a(i)             ! a is written to below
+                    ic3 = ic2                ! Read ic2 before write
+                    ic2 = 5                  ! write to ic2 is not first use
+                    ic4 = i-1                ! written twice
+                    if (i.eq.3) then         ! Nothing inside another statement
+                        ic5 = 1
+                    endif
+                    ic6 = i + some_func()    ! Function might depend on #calls
+                    a(ic1) = 1+(ic1+1)*ic1
+                    a(ic2) = 2+(ic2+1)*ic2
+                    a(ic3) = 3+(ic3+1)*ic3
+                    ic4 = i + 1              ! written twice
+                    a(ic4) = 4+(ic4+1)*ic4
+                    a(ic5) = 5+(ic5+1)*ic5
+                    a(ic6) = 6+(ic6+1)*ic6
+                end do
+                end program test'''
+    psyir = fortran_reader.psyir_from_source(source)
+    loop = psyir.children[0].children[0]
+
+    # None of the statements can be moved, so the output
+    # before and after the transformation should be identical:
+    out_before = fortran_writer(loop)
+    riv = ReplaceInductionVariables()
+    riv.apply(loop)
+    out_after = fortran_writer(loop)
+    assert out_before == out_after
+
+
+# ----------------------------------------------------------------------------
+def test_riv_other_step_size(fortran_reader, fortran_writer):
+    '''Tests loops with a step size different from one, which affects
+    the value to be assigned to replaced induction variables.
+    '''
+    source = '''program test
+                integer i, ic1, ic2
+                real, dimension(10) :: a
+                do i = 1, 10, 5
+                    ic1 = i+1
+                    ic2 = 3
+                    a(ic1) = 1+(ic1+1)*ic1
+                end do
+                end program test'''
+    psyir = fortran_reader.psyir_from_source(source)
+    loop = psyir.children[0].children[0]
+
+    riv = ReplaceInductionVariables()
+    riv.apply(loop)
+    out = fortran_writer(psyir)
+    assert "a(i + 1) = 1 + (i + 1 + 1) * (i + 1)" in out
+    assert "ic1 = i - 5 + 1" in out
+    assert "ic2 = 3" in out
+
+
+# ----------------------------------------------------------------------------
+@pytest.mark.parametrize("array_expr", ["ic(i)", "ic(2)", "t(i)%b", "t(1)%b"])
```

Comment:
For completeness can you add some SoA ("mys%a(i)" and "mys%b(i)%a") I did it myself and it works without any additional declaration because the unqualified import

Reply:
I've added more tests here.